### PR TITLE
Add jsdoc-eslint plugin

### PR DESCRIPTION
**Pull request by LuomaJuha**
Hello, i just thought if this plugin could bring some clarity for js.

https://www.npmjs.com/package/eslint-plugin-jsdoc

I left it as a draft for now as there is lots of writing to do and maybe something about what rules to enable etc but basically, it just makes sure that there is documentation in js code. Comments would be nice.

### CODE REVIEW

**Comment by demiankatz** [comment]
Thanks, @LuomaJuha, I think this is a good idea. Just a couple of initial thoughts:

1.) I know @crhallberg has already been thinking about introducing jsdoc in more places (e.g. to allow some Typescript-like type awareness without actually introducing Typescript). I'll be interested to hear from him whether anything here would conflict with his own plans, or if he has opinions on settings that should be adjusted.

2.) In terms of overall strategy, I think once we settle on the rules and format here, it may make more sense to incrementally separate open pull requests to add documentation to JS files rather than to try to do everything here all at once. That way, if we need to have discussions about particular functions, etc., the conversation thread here won't get too long. And obviously there's no harm in starting to merge more documentation into the code before we merge tooling to enforce the documentation. Once all the documentation has been merged, then we can merge this one last.

**Comment by LuomaJuha** [comment]
@demiankatz That sounds like a great strategy!

**Comment by LuomaJuha** [comment]
Made a small adjustment, where all the .js files except 2 are being ignored, if this approach would help with the introduction. It could that the pr could be merged before adjusting all the file so that there is no rush to change old files, but any new files would fall under the restrictions.

**Comment by demiankatz** [comment]
> Made a small adjustment, where all the .js files except 2 are being ignored, if this approach would help with the introduction. It could that the pr could be merged before adjusting all the file so that there is no rush to change old files, but any new files would fall under the restrictions.

@LuomaJuha, I would be happy to move forward with either strategy (merge documentation first, or exclude checks until documentation is added). If we go with the "exclude checks" strategy, though, I would still prefer to try to get all the documentation added as quickly as possible, since I'd rather not have these excludes in the long term, since they could eventually cause confusion or oversights.

If we do go forward with the excludes, I think it might make sense to use `**` instead of `bootstrap3` or `bootstrap5` in the patterns, since the bootstrap3 theme will be going away soon, and the filenames are all the same between those two themes anyway... it would nearly cut the configuration size in half.

In any case, I'd love to get some more opinions on this from others! This might be a good thing to highlight and discuss at the roadmapping session during the upcoming summit.

**Comment by crhallberg** [comment]
I am a big fan of the way this is being introduced! I've wanted to do this for future features but excluding files makes a clear target of todos.

1. ❓ If we're going to activate all of the recommended rules, would it make sense to either use the built-in recommended import:

    ```json
    {
      "extends": ["plugin:jsdoc/recommended"]
    }
    ```

    or sort them into an order where all the recommended and active rules are grouped:
    
    ```js
    {
      "rules": {
        // Recommended
        "jsdoc/check-access": 1, 
        "jsdoc/check-alignment": 1,
        ...

        // Disabled
        //"jsdoc/check-examples": 1,
        //"jsdoc/check-indentation": 1,
        //"jsdoc/check-line-alignment": 1,
        ...
      }
    ```
2. 👍 [I checked to see if we could target higher than es6 (aka es2015)](https://babeljs.io/docs/babel-plugin-transform-arrow-functions "check the left menu") but I think you nailed it. We don't have any features newer than that in the code.

**Comment by LuomaJuha** [comment]
@crhallberg Thanks for the comments! I left the bottom rule like that for now, so it is easier to leave comments, for which rules to apply, or just go with the default. I will try and make the adjustments today and open this for more comments.

File: build.xml [hunk 1]
```diff
@@ -274,6 +274,20 @@
     </exec>
   </target>
 
+  <!-- ESLint jsDoc -->
```
Comment (demiankatz) [inline build.xml]: Maybe it's worth differentiating these comments (i.e. add "(test only)" to this first one and "(auto-fix eligible issues)" to the second).

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the progress and discussion, @LuomaJuha and @crhallberg!

I have one minor suggestion below.

Beyond that, the next question is "what's the next step?" I think we agreed to start opening separate PRs to add documentation... but do we want to keep this one open and unmerged until all that work is done, or do we want to merge this and then uncomment exceptions as we fix them in separate PRs? Either solution seems workable to me, though if we merge this one, we should open a JIRA ticket to track the remaining work so we have a clear indicator that there is unfinished work in progress underway.

File: build.xml (hunk 1 above)
Comment (LuomaJuha) [reply build.xml]: Sure!

**Comment by LuomaJuha** [comment]
@demiankatz @crhallberg I think the plus thing about merging this now would be to avoid merge conflicts etc. Then create jira issue and open a subtask for the file or 2 - 3 files at a time? Just thoughts. After those are done, we could merge these configs into the original .eslintrc and unify it? Also, if we would add this to the qa-console, then any new files in the meanwhile would have to be documented?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Yes, you make good points @LuomaJuha. Do you want to add this to qa-console before we merge?

**Review by demiankatz — APPROVED** [review]
Thanks, @LuomaJuha!

**Comment by demiankatz** [comment]
I have opened [VUFIND-1717](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1717) to track the remaining work on this project.

**Pull request merged**